---
title: "Tumor Deconvolution"
author: "Anh Nguyet Vu"
output: html_document
date: "2024-06-06"
---

```{r setup, include=FALSE}

remotes::install_github("omnideconv/immunedeconv")
BiocManager::install("tximport")
BiocManager::install("ensembldb")
BiocManager::install("EnsDb.Hsapiens.v75")
```


```{r load, include=FALSE}

# Make sure packages load
library(tximport)
library(ensembldb)
library(immunedeconv)
library(EnsDb.Hsapiens.v75)
```

### Introduction

![tme-role-immune-cells](https://github.com/nf-osi/tutorials/assets/32753274/64c5a272-1c37-4df2-a91f-e1de71af4141)
 
In cancer biology, the "tumor microenvironment" (TME) comprises the immune cells, extracellular matrix, blood vessels and various additional tissue-resident cell types that are part of the tumor ecosystem [^1]. Here we are interested in one important component, the composition of different immune cell types. However, there isn't always direct data from more cell assays like immunohistochemistry, flow cytometry, and mass cytometry for quantifying immune infiltrates. For the JHU Biobank, for example, only sequencing data is available.

Fortunately, RNA-seq data can be used to characterize TME through a number of computational approaches. One highly used "tumor deconvolution" method is CIBERSORT[^2], which is available programmatically through R and a Python implementation via the TumorDecon package[^3]. (Other approaches are helpfully reviewed and benchmarked in [^4] and [^5].)


[^1]: https://doi.org/10.1016/j.ccell.2023.02.016
[^2]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5895181/
[^3]: https://doi.org/10.1016/j.softx.2022.101072
[^4]: https://academic.oup.com/bioinformatics/article/35/14/i436/5529146
[^5]: https://academic.oup.com/bib/article/22/4/bbaa219/5916936

### Locate files and clinical data

(Look Ma, no `synapser`!)

Files have already been transferred through the UI workflow and are just available under `/sbgenomics/project-files`.
```{r }

sf <- list.files("/sbgenomics/project-files", pattern = ".sf$", full.names = T) 
attributes <- read.csv("/sbgenomics/project-files/Job-3645258210835165589066307932.csv")
attributes

```

### Examining the expression data and mapping to gene-level expression

About these .sf files:
* They are quantification results at the transcript level.
* Each row corresponds to a transcript, with columns providing details such as the transcript ID, length, effective length, number of reads (TPM - transcripts per million), and estimated counts.
* Data in this file reflects the abundance of individual transcript isoforms.

Here's what one looks like -- just very tabular with columns as described above.
```{r sf-glimpse, echo=FALSE}

first_file <- sf[1]
read.csv(first_file, sep = "\t")
```

There can be multiple transcript isoforms per gene, and we actually need gene-level expression as inputs into the computational method.
These are Ensembl transcript IDs, the first one is http://useast.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000290825;r=1:11869-14409;t=ENST00000456328 

The input needs to be a matrix with genes in rows and samples in columns. The rownames must be HGNC symbols for human data.
In order to group/map these into gene-level summarization with HGNC gene symbols, we will use a mapping reference and the `tximport` package.

Since the mapping of raw reads was with Hg19 (see original paper), we will use Ensembl Release 75 to build the mapping table.

```{r gene-level, echo=FALSE}

edb <- EnsDb.Hsapiens.v75
tx <- transcripts(edb, columns=c("tx_name", "gene_name"), return.type = "data.frame")
tx2gene <- tx[, c("tx_name", "gene_name")]
head(tx2gene)

```

`tximport` conveniently reads all those sf files into the summary-level matrix.
```{r}
gx_matrices <- tximport(sf, type = "salmon", tx2gene = tx2gene, ignoreTxVersion = TRUE)
abundance <- gx_matrices$abundance
```


### Run CIBERSORT

The available methods for human data in `immunedeconv` are:
```
quantiseq
timer
cibersort
cibersort_abs
mcp_counter
xcell
epic
abis
consensus_tme
estimate
```

See https://github.com/omnideconv/immunedeconv for more about the methods implemented. 
Some popular methods are CIBERSORT and ESTIMATE (based on citations).
For our analysis, let's go with the recommendation as stated in their [benchmarking paper](https://academic.oup.com/bioinformatics/article/35/14/i436/5529146):

> we recommend EPIC and quanTIseq for general purpose deconvolution

The output is the estimated "absolute" fractions for each immune cell type.
```{r}

result <- immunedeconv::deconvolute(abundance, "quantiseq")
result

```

### Figures - initial

Of course, the results could be much more easily interpreted as a figure.
We'll plot what the immune cell distribution looks like for these cutaneous neurofibroma samples. 
Some studies such as have compared whether immune cell fractions vary over different types of cancers with solid tumors.
But cutaneous neurofibroma is relatively rare and unlikely to be included in these types of study comparisons.

```{r}

library(ggplot2)
library(reshape2)


data_long <- melt(result, varnames=c("ImmuneCellType", "Sample"), value.name="fraction")
ggplot(data_long, aes(x=cell_type, y=fraction, fill=cell_type)) +
  geom_boxplot() +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +  
  labs(title="Immune Cells in the Cutaneous Neurofibroma Tumor Environment",
       x="Immune Cell Type",
       y="Fraction") 
```

### Figures - extended

Well, some grouping might be interesting, and we should use the clinical attributes for the grouping.

```{r}

```

### Packages and versions

```{r message=TRUE, warning=FALSE}
sessionInfo()
```

### Citations
```{r}
citation("tximport")
citation("ensembldb")
citation("immunedeconv")
```
